### PR TITLE
Move transforms from 3d object detection dataset

### PR DESCRIPTION
**Pull request by kprokofi**
### Summary

Move 

- [x] Move affine transforms including resize from the dataset to torchvision transforms
- [x] Remove kitti utils
- [x] PIL -> OpenCV, align image ratio
- [x]  Fix OV IR Inference, make it works without OTX pre-processing 

### Checklist

- [x] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/core/model/detection_3d.py [hunk 1]
```diff
@@ -397,7 +423,7 @@ def _customize_inputs(
         self,
         entity: Det3DBatchDataEntity,
     ) -> dict[str, Any]:
-        img_sizes = np.array([img_info.ori_shape for img_info in entity.imgs_info])
+        img_sizes = np.array([img_info.ori_shape[::-1] for img_info in entity.imgs_info])
```
Comment (eunwoosh) [inline src/otx/core/model/detection_3d.py]: Is there any reason why setting img_size to WxH?

File: src/otx/core/data/dataset/object_detection_3d.py [hunk 2]
```diff
@@ -68,238 +55,55 @@ def __init__(
...
-            boxes_3d=torch.as_tensor(targets["boxes_3d"], dtype=torch.float32),
-            size_2d=torch.as_tensor(targets["size_2d"], dtype=torch.float32),
-            size_3d=torch.as_tensor(targets["size_3d"], dtype=torch.float32),
-            depth=torch.as_tensor(targets["depth"], dtype=torch.float32),
-            heading_angle=torch.as_tensor(
-                np.concatenate([targets["heading_bin"], targets["heading_res"]], axis=1),
-                dtype=torch.float32,
-            ),
+            boxes=np.zeros((self.max_objects, 4), dtype=np.float32),
+            labels=np.zeros((self.max_objects), dtype=np.int8),
```
Comment (eunwoosh) [inline src/otx/core/data/dataset/object_detection_3d.py:88]: `Det3DDataEntity` get torch but numpy as passed. If numpy needs to be passed, then could you update type hint of `Det3DDataEntity`? And could you revmoe `# mypy: ignore-errors` in this file?

File: src/otx/core/data/transform_libs/torchvision.py [hunk 3]
```diff
@@ -3449,6 +3449,304 @@ def __repr__(self) -> str:
         return repr_str
 
 
+class Decode3DInputsAffineTransforms(TopdownAffine):
```
Comment (eunwoosh) [inline src/otx/core/data/transform_libs/torchvision.py:3452]: Could you add unit test for this class?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work. I left some comments. Please take a look.

File: src/otx/core/model/detection_3d.py (hunk 1 above)
Comment (kprokofi) [reply src/otx/core/model/detection_3d.py:426]: I will change it to use HxW

Comment (kprokofi) [reply src/otx/core/model/detection_3d.py]: forward function and outputs post processing expect WxH image ratio. I will try to change this though

File: src/otx/core/data/dataset/object_detection_3d.py (hunk 2 above)
Comment (kprokofi) [reply src/otx/core/data/dataset/object_detection_3d.py:88]: Agree, but the problem that the decoding transform and all related functions expect numpy passed. However, the Det3DDataEntity intended to be with torch tensors for training. I will try to change this behavior somehow

File: src/otx/core/data/transform_libs/torchvision.py (hunk 3 above)
Comment (kprokofi) [reply src/otx/core/data/transform_libs/torchvision.py:3452]: Sure, will do

Comment (kprokofi) [reply src/otx/core/data/transform_libs/torchvision.py:3452]: Done https://github.com/openvinotoolkit/training_extensions/pull/4046/commits/ae60b71090bdc8f7e00776055de3f2c7e9684831

File: src/otx/core/data/dataset/object_detection_3d.py (hunk 2 above)
Comment (kprokofi) [reply src/otx/core/data/dataset/object_detection_3d.py:88]: Done https://github.com/openvinotoolkit/training_extensions/pull/4046/commits/ae60b71090bdc8f7e00776055de3f2c7e9684831

File: src/otx/core/model/detection_3d.py (hunk 1 above)
Comment (kprokofi) [reply src/otx/core/model/detection_3d.py]: Aligned to HxW everywhere https://github.com/openvinotoolkit/training_extensions/pull/4046/commits/ae60b71090bdc8f7e00776055de3f2c7e9684831

File: src/otx/core/model/detection_3d.py [hunk 4]
```diff
@@ -583,7 +609,7 @@ def _convert_pred_entity_to_compute_metric(
         .numpy()
     )
 
-    img_sizes = np.array([img_info.ori_shape for img_info in inputs.imgs_info])
+    img_sizes = np.array([img_info.ori_shape for img_info in inputs.imgs_info])  # HxW -> WxH
```
Comment (eunwoosh) [inline src/otx/core/model/detection_3d.py]:
```suggestion
    img_sizes = np.array([img_info.ori_shape for img_info in inputs.imgs_info])
```

**Review by eunwoosh — COMMENTED** [review]
Thanks for applying comment :) I left a minor comment. Could you check it?

File: src/otx/core/data/transform_libs/torchvision.py [hunk 5]
```diff
@@ -3449,6 +3449,333 @@ def __repr__(self) -> str:
...
+    @staticmethod
+    def rect_to_img(p2: np.ndarray, pts_rect: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
+        """Convert camera coordinates to image coordinates.
+
+        Args:
+            p2 (np.ndarray): Projection matrix with shape (3, 4).
+            pts_rect (np.ndarray): Rectangular coordinates with shape (N, 4).
+
+        Returns:
+            np.ndarray: Image coordinates with shape (N, 2).
```
Comment (sungchul1) [inline src/otx/core/data/transform_libs/torchvision.py]:
```suggestion
            tuple[np.ndarray, np.ndarray]: Image coordinates with shape (N, 2).
```

File: src/otx/core/data/transform_libs/torchvision.py (hunk 3 above)
Comment (sungchul1) [reply src/otx/core/data/transform_libs/torchvision.py:3452]: Nevertheless, most edge cases are not converted. Aren't unit tests needed to cover them?

File: src/otx/core/data/transform_libs/torchvision.py [hunk 6]
```diff
@@ -3449,6 +3449,333 @@ def __repr__(self) -> str:
...
+            msg = "angle not in 0 ~ 2pi"
+            raise ValueError(msg)
+
+        angle_per_class = 2 * np.pi / float(num_heading_bin)
+        shifted_angle = (angle + angle_per_class / 2) % (2 * np.pi)
+        class_id = int(shifted_angle / angle_per_class)
+        residual_angle = shifted_angle - (class_id * angle_per_class + angle_per_class / 2)
+        return class_id, residual_angle
+
+    def convert(self, inputs: T_OTXDataEntity, mask_2d: np.ndarray, image_to_tensor: bool = False) -> T_OTXDataEntity:  # type: ignore[override]
```
Comment (sungchul1) [inline src/otx/core/data/transform_libs/torchvision.py:3765]: Isn't it needed to be adapted in mixin for other transforms for 3d detection can be added in the future?
Or leaving a comment looks good.

File: src/otx/core/data/transform_libs/torchvision.py (hunk 3 above)
Comment (sungchul1) [reply src/otx/core/data/transform_libs/torchvision.py:3452]: And this transform seems to have multiple functionality, I think splitting it (decode and affine transform) looks better if possible. What do you think?

Comment (sungchul1) [reply src/otx/core/data/transform_libs/torchvision.py:3452]: And could you explain the reason for inheriting `TopdownAffine`? I didn't find the clue.
Why I ask it is that if `convert` is updated here and there is no reason to inherit `TopdownAffine`, it can inherit `tvt_v2.Transform` only.

File: src/otx/core/data/transform_libs/torchvision.py (hunk 6 above)
Comment (kprokofi) [reply src/otx/core/data/transform_libs/torchvision.py:3765]: This transform is used only for 3d detection task, I don't think it should be updated in mixin

File: src/otx/core/data/transform_libs/torchvision.py (hunk 3 above)
Comment (kprokofi) [reply src/otx/core/data/transform_libs/torchvision.py:3452]: This transform has multiple functionality because it is hard to divide transforms from decoding KITTI annotations for 3d detection training. The annotations should use info from affine transforms. I created this unified transform to move resizing, cropping and flipping from __get_item__ in the Dataset. 

This transform inherits from TopdownAffine, because it uses 2 its main methods, "_get_warp_matrix" and " _get_warp_image"

Comment (sungchul1) [reply src/otx/core/data/transform_libs/torchvision.py:3452]: Thanks for the explanation. If so, I agree with your decision.

Comment (kprokofi) [reply src/otx/core/data/transform_libs/torchvision.py:3452]: Added more unit tests: https://github.com/openvinotoolkit/training_extensions/pull/4046/commits/23d767dcc23942efb027e24f3872010542530abb

**Review by eunwoosh — APPROVED** [review]

**Review by sovrasov — APPROVED** [review]
